### ISSUE TITLE
Loader appears to be loading last instance of model for all in query

### ISSUE THREAD

**Issue by Ovyerus** [fantix/gino#480]
* GINO version: 0.8.2
* Python version: 3.7.2
* asyncpg version: 0.18.3
* aiocontextvars version: 0.2.1
* PostgreSQL version: 9.5.16

### Description

I'm attempting to load rows with corresponding users and their roles relating to the row. However it appears as if Gino is replacing previous rows with later rows loaded values, resulting in all rows sharing the same loaded values as the last row.

### What I Did
(simplified)

```py
import enum
import pprint

from gino import Gino

db = Gino()


class AuthorRole(enum.Enum):
    owner = 1
    unassigned = 2


class Base:
    id = db.Column(db.Integer(), primary_key=True)


class Mod(db.Model, Base):
    __tablename__ = "mod"

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

        self._authors = []
        self._owner = None

    @property
    def authors(self):
        return self._authors

    @authors.setter
    def authors(self, value: "User"):
        if hasattr(value, "role"):
            if value.role.role == AuthorRole.owner:
                self._owner = value

        self._authors.append(value)

    @property
    def owner(self):
        return self._owner

    def to_dict(self):
        return {
            "id": self.id,
            "authors": [a.to_dict() for a in self._authors],
            "owner": self._owner.to_dict(),
        }


class User(db.Model, Base):
    __tablename__ = "user"

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

        self._role = None
        self._role_name = None

    @property
    def role(self):
        return self._role

    @role.setter
    def role(self, value: "ModAuthor"):
        self._role = value
        self._role_name = value.role.name

    def to_dict(self):
        return {
            "id": self.id,
            "role": self._role_name,
            "__": self._role.to_dict(),  # for testing
        }


class ModAuthor(db.Model, Base):
    __tablename__ = "user_mod"

    role = db.Column(db.Enum(AuthorRole), default=AuthorRole.unassigned)
    user_id = db.Column(None, db.ForeignKey("user.id", ondelete="CASCADE"))
    mod_id = db.Column(None, db.ForeignKey("mod.id", ondelete="CASCADE"))


async def route():
    async with db.with_bind("postgresql://localhost/gino"):
        await db.gino.create_all()

        u1 = await User.create()
        u2 = await User.create()
        u3 = await User.create()
        u4 = await User.create()
        m1 = await Mod.create()
        m2 = await Mod.create()
        await ModAuthor.create(mod_id=m1.id, user_id=u1.id, role=AuthorRole.owner)
        await ModAuthor.create(mod_id=m1.id, user_id=u2.id)
        await ModAuthor.create(mod_id=m2.id, user_id=u3.id, role=AuthorRole.owner)
        await ModAuthor.create(mod_id=m2.id, user_id=u4.id)

        query = Mod.outerjoin(ModAuthor).outerjoin(User).select()

        loader = Mod.distinct(Mod.id).load(
            authors=User.distinct(User.id).load(role=ModAuthor.distinct(ModAuthor.id))
        )
        results = await query.gino.load(loader).all()
        pprint.pprint([m.to_dict() for m in results])

        await db.gino.drop_all()


if __name__ == "__main__":
    import asyncio

    asyncio.run(route())
```

### What I Expect to Happen

(`to_json`ed output)

```jsonc
[
    {
        "authors": [
            {
                "__": {
                    // ...
                    "id": "1",
                    "mod_id": "1", // Same ID as the mod it should be a part of
                },
                // ...
            },
            {
                "__": {
                    // ...
                    "id": "2",
                    "mod_id": "1",
                },
                // ...
            },
        ],
        "id": "1",
        // ...
    },
    {
        "authors": [
            {
                "__": {
                    // ...
                    "id": "3",
                    "mod_id": "2", // Different mod ID, as it should
                },
                // ...
            },
            {
                "__": {
                    // ...
                    "id": "4",
                    "mod_id": "2",
                },
                // ...
            },
        ],
        "id": "2",
        // ...
    },
]
```

### What Actually Happens

(`to_json`ed output)

```jsonc
[
    {
        "authors": [
            {
                "__": {
                    // ...
                    "id": "1",
                    "mod_id": "2", // Different mod's ID!!
                },
                // ...
            },
            {
                "__": {
                    // ...
                    "id": "2",
                    "mod_id": "2",
                },
                // ...
            },
        ],
        "id": "1",
        // ...
    },
    {
        "authors": [
            {
                "__": {
                    // ...
                    "id": "1", // Same ID as previous mod's authors
                    "mod_id": "2", // Correct ID, because it belongs to this mod.
                },
                // ...
            },
            {
                "__": {
                    // ...
                    "id": "2",
                    "mod_id": "2",
                },
                // ...
            },
        ],
        "id": "2",
        // ...
    },
]
```

**Comment by sr229** [comment]
Heya guys, we run Gino in production and this seems to be a major bug in our end as well, probably could use escalation.

**Comment by fantix** [comment]
Sorry for the inconvenience! I'll get to this once I have time.

**Comment by fantix** [comment]
I made some small changes in your demo (pls see above) to make it work for testing, and generated this:

```
[{'authors': [{'__': {'id': 1,
                      'mod_id': 1,
                      'role': <AuthorRole.owner: 1>,
                      'user_id': 1},
               'id': 1,
               'role': 'owner'},
              {'__': {'id': 2,
                      'mod_id': 1,
                      'role': <AuthorRole.unassigned: 2>,
                      'user_id': 2},
               'id': 2,
               'role': 'unassigned'}],
  'id': 1,
  'owner': {'__': {'id': 1,
                   'mod_id': 1,
                   'role': <AuthorRole.owner: 1>,
                   'user_id': 1},
            'id': 1,
            'role': 'owner'}},
 {'authors': [{'__': {'id': 3,
                      'mod_id': 2,
                      'role': <AuthorRole.owner: 1>,
                      'user_id': 3},
               'id': 3,
               'role': 'owner'},
              {'__': {'id': 4,
                      'mod_id': 2,
                      'role': <AuthorRole.unassigned: 2>,
                      'user_id': 4},
               'id': 4,
               'role': 'unassigned'}],
  'id': 2,
  'owner': {'__': {'id': 3,
                   'mod_id': 2,
                   'role': <AuthorRole.owner: 1>,
                   'user_id': 3},
            'id': 3,
            'role': 'owner'}}]
```

**Comment by fantix** [comment]
Please feel free to let me know if you meet further issues or bugs. 😃

**Comment by Ovyerus** [comment]
Aha, removing that `.distinct` did help. Wasn't sure exactly what that did, but I saw it used often in the loader examples so I thought it was important. Thanks for the help 👍 

As for `one Mod may have multiple owners, but the code only allows one`, it's kind of a leftover from before "roles", where everyone was an owner, then we dropped them in as a way to limit what some people can do. Probably is a good idea to add a check/separate column, will have to do some discussion.
